File: javascript/meetup/meetup.js
export const meetup = (year, month, descriptor, dayOfWeek) => {
	let dayMapping = { Sunday: 0, Monday: 1, Tuesday: 2, Wednesday: 3, Thursday: 4, Friday: 5, Saturday: 6 };
	let weekMapping = { first: 0, second: 1, third: 2, fourth: 3 };

	let dayParsed = dayMapping[dayOfWeek];
	let daysInCurrentMonth = daysInMonth(year, month);
	let teenth = [13, 14, 15, 16, 17, 18, 19];
	
	let dateFound = null;

	if (descriptor === "teenth") {
		let teenthDateDay = teenth.filter(day => new Date(year, month - 1, day).getDay() === dayParsed);

		dateFound = new Date(year, month - 1, teenthDateDay[0]);
	}
	else {
		let dateDays = [...Array(daysInCurrentMonth).keys()].map(x => x + 1).filter(day => new Date(year, month - 1, day).getDay() === dayParsed);
		
		dateFound = descriptor === "last" ? new Date(year, month - 1, dateDays[dateDays.length - 1]) : new Date(year, month - 1, dateDays[weekMapping[descriptor]]);;	
	}

	return dateFound;
};

function daysInMonth(year, month) {
	return new Date(year, month, 0).getDate();
}

File: csharp/meetup/Meetup.cs
using System;
using System.Collections.Generic;
using System.Linq;

public enum Schedule
{
    Teenth = 13,
    First = 1,
    Second = 8,
    Third = 15,
    Fourth = 22,
    Last
}

public class Meetup
{
    private readonly int _month;
    private readonly int _year;

    public Meetup(int month, int year)
    {
        _month = month;
        _year = year;
    }

    public DateTime Day(DayOfWeek dayOfWeek, Schedule schedule)
    {
        var daysInMonth = DateTime.DaysInMonth(_year, _month);
        DateTime startDate = new DateTime();

        startDate = new DateTime(_year, _month, (int)schedule);

        return schedule == Schedule.Last ? GetDateRange(new DateTime(_year, _month, (int)Schedule.Fourth), new DateTime(_year,_month, daysInMonth)).Where(x => x.DayOfWeek == dayOfWeek).Last() :
                                            GetDateRange(startDate, startDate.AddDays(6)).Where(x => x.DayOfWeek == dayOfWeek).Single();
    }

    private IEnumerable<DateTime> GetDateRange(DateTime startDate, DateTime endDate)
    {
        while(startDate <= endDate)
        {
            yield return startDate;
            startDate = startDate.AddDays(1);
        }
    }
}